tutorials/nservicebus-101/lesson-2/tutorial.md

```diff
@@ -0,0 +1,191 @@
+
+Also, you should not embed logic within your NServiceBus message classes. Each message should contain only automatic properties and not computed properties or methods. Also, it is a good practice to instantiate collection properties from a default parameterless constructor as shown above, so that you never have to deal with a potentially null collection.
+
+In essence, messages should be carriers for data only. This way, discovering a bug in logic will not require redeployment of every endpoint that uses a message.
+
+
+## How do I organize messages?
+
+Because messages must be shared between multiple endpoints, you can't put the classes in the same assembly with the endpoints. They need to live in a separate class library so that they can be referenced by multiple endpoints.
+
+A **message assembly** should contain only NServiceBus message contracts, and any supporting types required by the messages themselves. For example, a message may use an enumeration type for one of its properties; that enumeration should also live within the same message assembly.
+
+Message assemblies should be entirely self-contained. They should have no dependencies other than libraries included with the .NET Framework, and the NServiceBus core assembly, which is required to reference the `ICommand` interface. Limiting dependencies makes your message contracts more resilient to future changes in the system.
+
+NOTE: In another course, we'll learn how (and why) to use [message conventions](/nservicebus/messaging/conventions.md) to identify message types, rather than the `ICommand` interface. Then we don't even need to reference the NServiceBus assembly in our message assemblies. But, we'll leave that out of scope for now.
+
+
+## How do I process a message?
+
+To process a message, we create a **message handler**, a class that implements `IHandleMessages<T>`, where `T` is a message type. A message handler looks like this:
+
+snippet:EmptyHandler
+
+The implementation of the `IHandleMessages<T>` interface is the `Handle` method, which NServiceBus will invoke when a matching message (in this case `DoSomething`) arrives. The `Handle` method receives the message and an `IMessageHandlerContext` that contains tools for working with messages.
+
+The `Handle` method must return a [`System.Threading.Tasks.Task`](https://msdn.microsoft.com/en-us/library/system.threading.tasks.task.aspx). Returning `null` will result in an exception. In the case above, no async calls need to be made, so we return `Task.CompletedTask` to fulfill the return requirement. 
+
+NOTE: Before .NET 4.6, `Task.CompletedTask` is not available. In that case, you can use `Task.FromResult(false)` instead.
+
+Instead, we can add the `async` keyword to the method, and then the compiler takes care of the tasks more or less transparently. No explicit return is needed because the compiler generates it for you as part of the async state machine.
```

Comment:
i think remove the above 3 paras and instead add a link here https://docs.particular.net/nservicebus/handlers/async-handlers